src/VisualStudio/Core/Def/Implementation/Diagnostics/DiagnosticProgressReporter.cs

```diff
@@ -0,0 +1,139 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.ComponentModel.Composition;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.SolutionCrawler;
+using Microsoft.VisualStudio.Shell;
+using Microsoft.VisualStudio.TaskStatusCenter;
+
+namespace Microsoft.VisualStudio.LanguageServices.Implementation.Diagnostics
+{
+    [Export(typeof(DiagnosticProgressReporter))]
+    internal class DiagnosticProgressReporter
+    {
+        private readonly IVsTaskStatusCenterService _taskCenterService;
+        private readonly IDiagnosticService _diagnosticService;
+        private readonly TaskHandlerOptions _options;
+
+        // this fields are never accessed concurrently
+        private TaskCompletionSource<bool> _currentTask;
+        private string _lastMessage;
+
+        // this is only field that is shared between 2 events streams (IDiagnosticService and ISolutionCrawlerProgressReporter)
+        // and can be called concurrently.
+        // volatile to make sure value is shared properly between threads
+        private volatile ITaskHandler _taskHandler;
+
+        [ImportingConstructor]
+        public DiagnosticProgressReporter(
+            SVsServiceProvider serviceProvider,
+            IDiagnosticService diagnosticService,
+            VisualStudioWorkspace workspace)
+        {
+            // no event unsubscription since it will remain alive until VS shutdown
+            _taskCenterService = (IVsTaskStatusCenterService)serviceProvider.GetService(typeof(SVsTaskStatusCenterService));
+            _diagnosticService = diagnosticService;
+
+            var crawlerService = workspace.Services.GetService<ISolutionCrawlerService>();
+            var reporter = crawlerService.GetProgressReporter(workspace);
+
+            Started(reporter.InProgress);
+
+            reporter.ProgressChanged += OnSolutionCrawlerProgressChanged;
+            _diagnosticService.DiagnosticsUpdated += OnDiagnosticsUpdated;
+
+            _options = new TaskHandlerOptions()
+            {
+                Title = ServicesVSResources.Live_code_analysis,
+                ActionsAfterCompletion = CompletionActions.None
+            };
+        }
+
+        private void OnDiagnosticsUpdated(object sender, DiagnosticsUpdatedArgs e)
+        {
+            // there is no concurrent call to this method since IDiagnosticService will serialize all
+            // events to preserve event ordering
```

Comment:
❗️ Document this constraint explicitly on the `IDiagnosticService.DiagnosticsUpdated` event, or else we cannot safely assume it here.

Reply:
sure